File: solutions/0003-longest-substring-without-repeating-characters/Solution.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var res = 0

        var start = 0
        val lastIndex = HashMap<Char, Int>()
        for ((end, c) in s.withIndex()) {
            start = Math.max(start, lastIndex.getOrDefault(c, -1) + 1)
            lastIndex[c] = end
            res = Math.max(res, end - start + 1)
        }

        return res
    }
}

File: solutions/0003-longest-substring-without-repeating-characters/solution.ts
function lengthOfLongestSubstring(s: string): number {
  let res = 0;

  const lastIndex = new Map<string, number>();
  let start = 0;
  for (let end = 0; end < s.length; ++end) {
    const c = s[end];
    start = Math.max(start, (lastIndex.get(c) ?? -1) + 1);
    lastIndex.set(c, end);
    res = Math.max(res, end - start + 1);
  }

  return res;
}
